pkg/webhook/webhook.go

```diff
@@ -0,0 +1,104 @@
+	KFServingWebhookServerNamespace = "kubeflow-system"
+)
+var log = logf.Log.WithName(KFServingWebhookServerName)
+
+// AddToManager adds all Controllers to the Manager
+// +kubebuilder:rbac:groups=admissionregistration.k8s.io,resources=mutatingwebhookconfigurations;validatingwebhookconfigurations,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups="",resources=secrets,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups="",resources=services,verbs=get;list;watch;create;update;patch;delete
+func AddToManager(manager manager.Manager) error {
+	server, err := webhook.NewServer("webhook-server", manager, webhook.ServerOptions{
+		Port:    9876,
+		CertDir: "/tmp/cert",
+		BootstrapOptions: &webhook.BootstrapOptions{
+			Secret: &types.NamespacedName{
+				Namespace: KFServingWebhookServerNamespace,
+				Name:      KFServingWebhookServerName + "-secret",
+			},
+			Service: &webhook.Service{
+				Namespace: KFServingWebhookServerNamespace,
+				Name:      KFServingWebhookServerName + "-service",
+				Selectors: map[string]string{
+					"control-plane": "controller-manager",
+				},
+			},
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	if err := register(manager, server); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// In 1.13, replace with the following webhook generators: ValidatingWebhookFor, DefaultingWebhookFor
+// https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/webhook/admission/validator.go#L35
+// https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/webhook/admission/defaulter.go#L34
+func register(manager manager.Manager, server *webhook.Server) error {
+	return server.Register(&admission.Webhook{
+		Name: KFServingWebhookServerName + serving.APIGroupName + "-validator",
+		Type: webooktypes.WebhookTypeValidating,
+		Rules: []v1beta1.RuleWithOperations{{
+			Operations: []v1beta1.OperationType{
+				v1beta1.Create,
+				v1beta1.Update,
+			},
+			Rule: v1beta1.Rule{
+				APIGroups:   []string{serving.APIGroupName},
+				APIVersions: []string{v1alpha1.APIVersion},
+				Resources:   []string{v1alpha1.KFServiceAPIName},
+			},
+		}},
+		Handlers: []admission.Handler{
+			&kfservice.Validator{
+				Client:  manager.GetClient(),
+				Decoder: manager.GetAdmissionDecoder(),
+			},
+		},
+	}, &admission.Webhook{
+		Name: KFServingWebhookServerName + serving.APIGroupName + "-defaulter",
+		Type: webooktypes.WebhookTypeMutating,
+		Rules: []v1beta1.RuleWithOperations{{
+			Operations: []v1beta1.OperationType{
+				v1beta1.Create,
+				v1beta1.Update,
+			},
+			Rule: v1beta1.Rule{
+				APIGroups:   []string{serving.APIGroupName},
+				APIVersions: []string{v1alpha1.APIVersion},
+				Resources:   []string{v1alpha1.KFServiceAPIName},
+			},
+		}},
+		Handlers: []admission.Handler{
+			&kfservice.Defaulter{
+				Client:  manager.GetClient(),
+				Decoder: manager.GetAdmissionDecoder(),
+			},
+		},
```

Comment:
we should setup `FailurePolicy` for the webhook, by default it is `Ignore` but we want `Fail`.

Reply:
done.